File: src/backend/langgraph/nodes/applyWithRollback.ts
import type { Tool, ToolProgress } from "src/Tool";
import type { LoopContext } from "src/loopContext";
import { captureKnownGood, planRollback } from "src/state/rollback";
import { applyStack } from "src/tools/applyStack";
import { destroyStack } from "src/tools/destroyStack";
import {
  type ConfigFileSnapshot,
  type StagedConfigFile,
  restoreConfigFiles,
  snapshotConfigFiles,
  writeConfigFiles,
} from "src/tools/shared/configFiles";
import type { LoopEvent } from "src/types/events";

export interface ApplyWithRollbackParams {
  stackName: string;
  desiredYaml: string;
  scaleOverrides?: Record<string, number>;
  configFiles: StagedConfigFile[];
  ctx: LoopContext;
  emit: (ev: LoopEvent) => void;
}

export interface ApplyWithRollbackResult {
  ok: boolean;
  resultMessage: string;
}

async function runApplyTool<TIn, TOut>(
  tool: Tool<TIn, TOut>,
  input: TIn,
  ctx: LoopContext,
  emit: (ev: LoopEvent) => void,
): Promise<TOut> {
  emit({ type: "tool_call", name: tool.name, input });
  const gen = tool.call(input, ctx);
  let result: TOut;
  while (true) {
    const r = await gen.next();
    if (r.done) {
      result = r.value;
      break;
    }
    emit({ type: "tool_progress", msg: (r.value as ToolProgress).msg });
  }
  emit({ type: "tool_result", name: tool.name, output: result });
  return result;
}

export async function runApplyWithRollback(
  params: ApplyWithRollbackParams,
): Promise<ApplyWithRollbackResult> {
  const { stackName, desiredYaml, scaleOverrides, configFiles, ctx, emit } = params;

  // Capture known-good state BEFORE applyStack overwrites on-disk state
  const known = captureKnownGood(stackName, ctx);

  // Write the agent-authored config files referenced by bind mounts, snapshotting
  // prior state so a failed apply can be rolled back. Confined to ctx.cwd.
  const configSnapshots: ConfigFileSnapshot[] = snapshotConfigFiles(ctx.cwd, configFiles);
  try {
    writeConfigFiles(ctx.cwd, configFiles);
  } catch (err) {
    restoreConfigFiles(configSnapshots);
    return {
      ok: false,
      resultMessage: `failed to write config files: ${(err as Error).message}`,
    };
  }

  const applyInput = applyStack.inputSchema.parse({
    stackName,
    composeYaml: desiredYaml,
    ...(scaleOverrides && Object.keys(scaleOverrides).length ? { scaleOverrides } : {}),
  });
  const applyResult = await runApplyTool(applyStack, applyInput, ctx, emit);

  if (applyResult.ok) {
    return { ok: true, resultMessage: "Stack applied." };
  }

  // Determine failure reason
  const reason = applyResult.healthy === false ? "unhealthy" : "apply_failed";
  const logsTail = applyResult.errorOutput ? `\nRecent logs:\n${applyResult.errorOutput}` : "";
  const detail =
    reason === "unhealthy"
      ? `unhealthy: ${(applyResult.unhealthyServices ?? []).join(", ")}${logsTail}`
      : `exit ${applyResult.exitCode}: ${applyResult.errorOutput ?? "unknown"}`;

  emit({
    type: "rollback_started",
    stackName,
    reason,
    detail,
    ...(applyResult.runningServices ? { runningServices: applyResult.runningServices } : {}),
  });

  const plan = planRollback(known, stackName);
  let restored: "previous" | "removed" | "none" = "none";
  let rollbackOk = true;

  try {
    if (plan.strategy === "restore_previous") {
      // UPDATE with recoverable prior -> re-apply it
      const restoreInput = applyStack.inputSchema.parse({
        stackName,
        composeYaml: plan.composeYaml,
      });
      const restore = await runApplyTool(applyStack, restoreInput, ctx, emit);
      rollbackOk = restore.ok;
      restored = "previous";
    } else if (plan.strategy === "teardown_partial") {
      // FIRST-TIME CREATE -> tear down partial stack
      const downInput = destroyStack.inputSchema.parse({ stackName });
      const down = await runApplyTool(destroyStack, downInput, ctx, emit);
      rollbackOk = down.ok;
      restored = "removed";
    } else {
      // UPDATE expected but unrecoverable -> abort, do NOT modify on-disk state
      rollbackOk = false;
      restored = "none";
    }
  } catch {
    rollbackOk = false;
  }

  restoreConfigFiles(configSnapshots);

  ctx.stateStore.appendHistory({
    ts: new Date().toISOString(),
    sessionId: ctx.sessionId ?? "unknown",
    stackName,
    action: "rollback",
    details: { reason, restored, rollbackOk },
  });

  emit({
    type: "rollback_result",
    stackName,
    ok: rollbackOk,
    restored,
    ...(!rollbackOk ? { detail: "manual intervention may be required" } : {}),
  });

  return {
    ok: false,
    resultMessage: `apply failed (${detail}); rollback ${rollbackOk ? "succeeded" : "FAILED"} (${restored}).`,
  };
}

File: src/engine/nodes/apply_with_rollback.py
"""Apply a desired stack YAML with rollback on failure.

Parity: ``src/backend/langgraph/nodes/applyWithRollback.ts``.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any, Literal

from src.engine.adapters.tool_adapter import run_tool
from src.state.rollback import capture_known_good, plan_rollback
from src.state.state_store import HistoryEvent
from src.tools.apply_stack import apply_stack
from src.tools.destroy_stack import destroy_stack
from src.tools.shared.config_files import (
    StagedConfigFile,
    restore_config_files,
    snapshot_config_files,
    write_config_files,
)
from src.types.events import (
    RollbackResult,
    RollbackStarted,
    ToolCall,
    ToolProgress,
    ToolResult,
)


@dataclass
class ApplyWithRollbackParams:
    stack_name: str
    desired_yaml: str
    config_files: list[StagedConfigFile]
    ctx: Any
    emit: Callable[[Any], None]
    scale_overrides: dict[str, int] | None = None


@dataclass
class ApplyWithRollbackResult:
    ok: bool
    result_message: str


async def _run_apply_tool(
    tool: Any,
    input_data: Any,
    ctx: Any,
    emit: Callable[[Any], None],
) -> Any:
    emit(ToolCall(name=tool.name, input=input_data))
    run = await run_tool(tool, input_data, ctx)
    for p in run.progress:
        emit(ToolProgress(msg=p.msg))
    emit(ToolResult(name=tool.name, output=run.output))
    return run.output


async def run_apply_with_rollback(params: ApplyWithRollbackParams) -> ApplyWithRollbackResult:
    stack_name = params.stack_name
    ctx = params.ctx
    emit = params.emit

    known = capture_known_good(stack_name, {"state_store": ctx.state_store})
    snapshots = snapshot_config_files(ctx.cwd, params.config_files)
    try:
        write_config_files(ctx.cwd, params.config_files)
    except Exception as err:
        restore_config_files(snapshots)
        return ApplyWithRollbackResult(
            ok=False, result_message=f"failed to write config files: {err}"
        )

    apply_input = apply_stack.input_schema.model_validate(
        {
            "stack_name": stack_name,
            "compose_yaml": params.desired_yaml,
            **(
                {"scale_overrides": params.scale_overrides}
                if params.scale_overrides
                else {}
            ),
        }
    )
    apply_result = await _run_apply_tool(apply_stack, apply_input, ctx, emit)

    if apply_result.ok:
        return ApplyWithRollbackResult(ok=True, result_message="Stack applied.")

    reason = "unhealthy" if apply_result.healthy is False else "apply_failed"
    logs_tail = f"\nRecent logs:\n{apply_result.error_output}" if apply_result.error_output else ""
    detail = (
        f"unhealthy: {', '.join(apply_result.unhealthy_services or [])}{logs_tail}"
        if reason == "unhealthy"
        else f"exit {apply_result.exit_code}: {apply_result.error_output or 'unknown'}"
    )

    emit(
        RollbackStarted(
            stackName=stack_name,
            reason=reason,
            detail=detail,
            runningServices=apply_result.running_services,
        )
    )

    plan = plan_rollback(known, stack_name)
    restored: Literal["previous", "removed", "none"] = "none"
    rollback_ok = True

    try:
        if plan.strategy == "restore_previous":
            restore_input = apply_stack.input_schema.model_validate(
                {"stack_name": stack_name, "compose_yaml": plan.compose_yaml}
            )
            restore = await _run_apply_tool(apply_stack, restore_input, ctx, emit)
            rollback_ok = restore.ok
            restored = "previous"
        elif plan.strategy == "teardown_partial":
            down_input = destroy_stack.input_schema.model_validate({"stack_name": stack_name})
            down = await _run_apply_tool(destroy_stack, down_input, ctx, emit)
            rollback_ok = down.ok
            restored = "removed"
        else:
            rollback_ok = False
            restored = "none"
    except Exception:
        rollback_ok = False

    restore_config_files(snapshots)

    ctx.state_store.append_history(
        HistoryEvent(
            ts=datetime.now(UTC).isoformat(),
            session_id=ctx.session_id or "unknown",
            stack_name=stack_name,
            action="rollback",
            details={"reason": reason, "restored": restored, "rollback_ok": rollback_ok},
        )
    )

    emit(
        RollbackResult(
            stackName=stack_name,
            ok=rollback_ok,
            restored=restored,
            **({"detail": "manual intervention may be required"} if not rollback_ok else {}),
        )
    )

    status = "succeeded" if rollback_ok else "FAILED"
    return ApplyWithRollbackResult(
        ok=False,
        result_message=f"apply failed ({detail}); rollback {status} ({restored}).",
    )